File: src/lib/api/gitWorkspace.ts
import { devApiFetch, hasTauriRuntime, invoke } from "./client";
import type {
  GitWorkspace,
  GitWorkspaceBranch,
  GitWorkspaceDetail,
  GitWorkspaceScanJobStatus,
  GitWorkspaceScanStartResult,
  AiCommitGitWorkspaceInput,
  AiCommitGitWorkspaceResult,
  ListGitWorkspacesInput,
  MergeGitWorkspaceBranchInput,
  ScanGitWorkspaceRootInput,
  ScanGitWorkspaceRootResult,
  SwitchGitWorkspaceBranchInput,
  UpsertGitWorkspaceInput,
} from "@/types";

function requireTauriRuntime(): never {
  throw new Error("Git 工作区需要读取本地仓库目录，请在 Tauri 桌面端使用该功能。");
}

export const gitWorkspaceApi = {
  list: (input?: ListGitWorkspacesInput) =>
    hasTauriRuntime()
      ? invoke<GitWorkspace[]>("list_git_workspaces", { input })
      : devApiFetch<GitWorkspace[]>("/git-workspaces/list", {
          method: "POST",
          body: JSON.stringify(input ?? null),
        }).catch(() => []),
  upsert: (input: UpsertGitWorkspaceInput) =>
    hasTauriRuntime()
      ? invoke<GitWorkspace>("upsert_git_workspace", { input })
      : Promise.resolve(requireTauriRuntime()),
  delete: (workspaceKey: string) =>
    hasTauriRuntime()
      ? invoke<void>("delete_git_workspace", { workspaceKey })
      : Promise.resolve(requireTauriRuntime()),
  refresh: (workspaceKey: string) =>
    hasTauriRuntime()
      ? invoke<GitWorkspace>("refresh_git_workspace", { workspaceKey })
      : Promise.resolve(requireTauriRuntime()),
  detail: (workspaceKey: string) =>
    hasTauriRuntime()
      ? invoke<GitWorkspaceDetail>("get_git_workspace_detail", { workspaceKey })
      : Promise.resolve(requireTauriRuntime()),
  scanRoot: (input: ScanGitWorkspaceRootInput) =>
    hasTauriRuntime()
      ? invoke<ScanGitWorkspaceRootResult>("scan_git_workspace_root", { input })
      : Promise.resolve(requireTauriRuntime()),
  startScanRoot: (input: ScanGitWorkspaceRootInput) =>
    hasTauriRuntime()
      ? invoke<GitWorkspaceScanStartResult>("start_git_workspace_root_scan", { input })
      : Promise.resolve(requireTauriRuntime()),
  getScanStatus: (jobId: string) =>
    hasTauriRuntime()
      ? invoke<GitWorkspaceScanJobStatus>("get_git_workspace_scan_status", { jobId })
      : Promise.resolve(requireTauriRuntime()),
  aiCommit: (input: AiCommitGitWorkspaceInput) =>
    hasTauriRuntime()
      ? invoke<AiCommitGitWorkspaceResult>("ai_commit_git_workspace", { input })
      : Promise.resolve(requireTauriRuntime()),
  pull: (workspaceKey: string) =>
    hasTauriRuntime()
      ? invoke<GitWorkspace>("pull_git_workspace", { workspaceKey })
      : Promise.resolve(requireTauriRuntime()),
  push: (workspaceKey: string) =>
    hasTauriRuntime()
      ? invoke<GitWorkspace>("push_git_workspace", { workspaceKey })
      : Promise.resolve(requireTauriRuntime()),
  branches: (workspaceKey: string) =>
    hasTauriRuntime()
      ? invoke<GitWorkspaceBranch[]>("list_git_workspace_branches", { workspaceKey })
      : Promise.resolve(requireTauriRuntime()),
  switchBranch: (input: SwitchGitWorkspaceBranchInput) =>
    hasTauriRuntime()
      ? invoke<GitWorkspace>("switch_git_workspace_branch", { input })
      : Promise.resolve(requireTauriRuntime()),
  mergeBranch: (input: MergeGitWorkspaceBranchInput) =>
    hasTauriRuntime()
      ? invoke<GitWorkspace>("merge_git_workspace_branch", { input })
      : Promise.resolve(requireTauriRuntime()),
};

File: src-tauri/src/commands/git_workspace.rs
use crate::error::CommandError;
use crate::models::{
    AiCommitGitWorkspaceInput, AiCommitGitWorkspaceResult, GitWorkspace, GitWorkspaceBranch,
    GitWorkspaceDetail, GitWorkspaceScanJobStatus, GitWorkspaceScanStartResult,
    ListGitWorkspacesInput, ScanGitWorkspaceRootInput, ScanGitWorkspaceRootResult,
    SwitchGitWorkspaceBranchInput, UpsertGitWorkspaceInput,
};
use crate::services::git_workspace::GitWorkspaceService;
use crate::state::AppState;

#[tauri::command]
pub fn list_git_workspaces(
    state: tauri::State<'_, AppState>,
    input: Option<ListGitWorkspacesInput>,
) -> Result<Vec<GitWorkspace>, CommandError> {
    GitWorkspaceService::list(&state.db, input).map_err(|e| e.into())
}

#[tauri::command]
pub async fn upsert_git_workspace(
    state: tauri::State<'_, AppState>,
    input: UpsertGitWorkspaceInput,
) -> Result<GitWorkspace, CommandError> {
    GitWorkspaceService::upsert(&state.db, input)
        .await
        .map_err(|e| e.into())
}

#[tauri::command]
pub fn delete_git_workspace(
    state: tauri::State<'_, AppState>,
    workspace_key: String,
) -> Result<(), CommandError> {
    GitWorkspaceService::delete(&state.db, &workspace_key).map_err(|e| e.into())
}

#[tauri::command]
pub async fn refresh_git_workspace(
    state: tauri::State<'_, AppState>,
    workspace_key: String,
) -> Result<GitWorkspace, CommandError> {
    GitWorkspaceService::refresh(&state.db, &workspace_key)
        .await
        .map_err(|e| e.into())
}

#[tauri::command]
pub async fn get_git_workspace_detail(
    state: tauri::State<'_, AppState>,
    workspace_key: String,
) -> Result<GitWorkspaceDetail, CommandError> {
    GitWorkspaceService::detail(&state.db, &workspace_key)
        .await
        .map_err(|e| e.into())
}

#[tauri::command]
pub async fn scan_git_workspace_root(
    state: tauri::State<'_, AppState>,
    input: ScanGitWorkspaceRootInput,
) -> Result<ScanGitWorkspaceRootResult, CommandError> {
    GitWorkspaceService::scan_root(&state.db, input)
        .await
        .map_err(|e| e.into())
}

#[tauri::command]
pub fn start_git_workspace_root_scan(
    app: tauri::AppHandle,
    input: ScanGitWorkspaceRootInput,
) -> Result<GitWorkspaceScanStartResult, CommandError> {
    GitWorkspaceService::start_scan_root(app, input).map_err(|e| e.into())
}

#[tauri::command]
pub fn get_git_workspace_scan_status(
    job_id: String,
) -> Result<GitWorkspaceScanJobStatus, CommandError> {
    GitWorkspaceService::get_scan_status(&job_id).map_err(|e| e.into())
}

#[tauri::command]
pub async fn ai_commit_git_workspace(
    state: tauri::State<'_, AppState>,
    input: AiCommitGitWorkspaceInput,
) -> Result<AiCommitGitWorkspaceResult, CommandError> {
    GitWorkspaceService::ai_commit(&state.db, input)
        .await
        .map_err(|e| e.into())
}

#[tauri::command]
pub async fn pull_git_workspace(
    state: tauri::State<'_, AppState>,
    workspace_key: String,
) -> Result<GitWorkspace, CommandError> {
    GitWorkspaceService::pull(&state.db, &workspace_key)
        .await
        .map_err(|e| e.into())
}

#[tauri::command]
pub async fn list_git_workspace_branches(
    state: tauri::State<'_, AppState>,
    workspace_key: String,
) -> Result<Vec<GitWorkspaceBranch>, CommandError> {
    GitWorkspaceService::branches(&state.db, &workspace_key)
        .await
        .map_err(|e| e.into())
}

#[tauri::command]
pub async fn switch_git_workspace_branch(
    state: tauri::State<'_, AppState>,
    input: SwitchGitWorkspaceBranchInput,
) -> Result<GitWorkspace, CommandError> {
    GitWorkspaceService::switch_branch(&state.db, input)
        .await
        .map_err(|e| e.into())
}
